Pass on the Layout chapter (#12)
pavpanchekha/emberfox: book/layout.md

modes compute sizes and positions differently.

The inline layout mode is effectively what we implemented in [Chapter
3](text.md), so this chapter will mostly be implementing block layout.

Second: the layout tree. Tree-based layout starts by recursively
traversing the HTML tree to build the layout tree. With two layout
modes, the layout objects will have one of two types: inline and
block.

We'll follow a simple rule, which is that a block layout either
contains any number of blocks, or a single inline, while inline layout
has no children. So for example, a document with a heading and two
paragraphs would correspond to a tree like this:

[^in-english]: In European languages. But not universally!

![A tree of layout objects. The root is a block for the `<html>`
    element; it has one child, a block for the `<body>` element; it
    has three children, each blocks as well (for the `<h1>` and two
    `<p>` elements); and each of those have a single child, an
    inline.](/im/layout-modes.png)

Finally: layout. With the layout tree created, how do we compute the
size and position of each layout object? The general rule is that a

> block, like a paragraph, should take up as much horizontal room as it

Reviewer: say "greedy" somewhere in here

can, and should be tall enough to contain everything inside it. That
means that a layout objects's width is based on its *parent*'s width,
while its *height* is based on its *children*'s height:

![The flow of information through layout. Width information flows down
    the tree, from parent to child, while height information flows up,
    from child to parent.](/im/layout-order.png)

This suggests a step-by-step approach to layout. First, an element

> must compute its width, based on its parent's width. That makes it

Reviewer: .. and its own style, which may override the width

possible to lay out the children, which comes next. Finally, the
children's heights are now available, so the element's height can be

> calculated.

Reviewer: Before python, I suggest some pseudocode:

```
Layout(parent_width) {
  my_width = ComputeWidth(parent_width, my_style)
  my_childrens_height = sum of Layout(children)
  my_height = ComputeHeight(my_childrens_height, my_style)
  return my_height
}

ComputeWidth(parent_width, my_style) {
  if (my_style prescribes a width)
   return width from my_style
  else
   return parent_width
}

ComputeHeight(my_childrens_height, my_style) {
  if (my_style prescribes a height)
    return height from my_style
  else
    return my_childrens_height
```

Let's now turn these concepts into code.

::: {.further}
Formally, computations on a tree like this can be
described by an [attribute grammar](wiki-atgram). Attribute grammar
engines analyze dependencies between different attributes to determine
the right order to calculate each attribute.
:::

[wiki-atgram]: https://en.wikipedia.org/wiki/Attribute_grammar

Layout modes
============

Let's start with the scaffolding for layout modes. We'll have two,
block and inline layout, and since what we've written now is basically
inline layout let's rename `Layout` to `InlineLayout` and add a new
`BlockLayout` class:

``` {.python}
class InlineLayout:
    # ...

class BlockLayout:
    # ...
```

These layout objects will be the nodes of the layout tree, and in fact

> some browsers call it a layout tree. To make it a tree, we'll want

Reviewer: Remove this second part of the sentence? Not sure it adds value and avoids confusion.

both types of layout objects to know their children, their parent, and
the HTML element they correspond to:

``` {.python}
def __init__(self, node, parent):
    self.node = node
    self.parent = parent
    self.children = []
    # ...
```

The `InlineLayout` constructor also initializes `weight`, `style`, and
`size`, as well as `x` and `y`, and it calls `recurse`. Let's move all
that into a `layout` method.

``` {.python expected=False}
def layout(self):
    self.display_list = []

    self.x = HSTEP
    self.y = VSTEP
    self.weight = "normal"
    self.style = "roman"
    self.size = 16

    self.line = []
    self.recurse(self.node)
    self.flush()
```

Besides a node, a parent, and children, each layout object also needs
a size and position, which we'll store in the `w`, `h`, `x`, and `y`
fields:

``` {.python}
def __init__(self, node, parent):
    # ...
    self.x = -1
    self.y = -1

>     self.w = -1

Reviewer: s/w/width/

s/h/height/

    self.h = -1
```

Unfortunately, `InlineLayout` already uses the `x` and `y` fields for
the location of the next word. To avoid a *very* annoying conflict,

> let's rename those fields to `cx` and `cy`. Instead of initializing

Reviewer: Choose something better than cx and cy. block_x_position?

4. An `InlineLayout` doesn't have children.

Size and position
=================

The layout tree is now made, but the size and position of each layout
object is still left unset. Let's fix that. The general strategy is
clear: each layout object computes its width, then lays out its
children, and then computes its height.

Besides width and height, we also need to position each element. This
is trickier, because if you have several paragraphs, the position of
the second depends on the height of the first. I'll use the rule that
each element it positioned by its parent before its own `layout`
method is called.

Let's start in `BlockLayout`. For the width, the idea is that a
paragraph or header or something like that should take up all the
horizontal space it can. So it should use the parent's width:

``` {.python}
self.w = self.parent.w
```

Then, each of the children need to be laid out. That means each child
has to be positioned, and its `layout` method must be called:

``` {.python indent=8}
y = self.y
for child in self.children:
    child.x = self.x
    child.y = y
    child.layout()
    y += child.h
```

Note that the call to the child's `layout` method not only asks it to

> compute its size, but also to build more of the layout tree.

Reviewer: Build more fo the tree? Or recursively compute its layout?

The height of an element must include all its children. Since the
children's heights have already all been summed into `y`, we can just
subtract its starting value to compute the height:

``` {.python}
self.h = y - self.y
```

That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`
and `y` will be set by its parent, so its `layout` just needs to set
`w` and `h`:

``` {.python}
class InlineLayout:
    def layout(self):
        self.w = self.parent.w
        # ...
        self.h = self.cy - self.y
```

Finally even `DocumentLayout` needs some layout code, though since the
document always starts in the same place it's pretty simple:

``` {.python}
class DocumentLayout:
    def layout(self):
        self.w = WIDTH
        # ...
        child.x = self.x = 0
        child.y = self.y = 0
        child.layout()

>         self.h = child.h

Reviewer: Or the height of the viewport instead. This makes it easier to explain how backgrounds of the root elements work, as well as clipping and scrolling of the "DocumentLayout".

3. Next, the object must lay out its children, which requires setting
   their `x` and `y` fields and calling their `layout` methods.
4. Finally, `layout` must set the object's `h` field.

Using tree-based layout
=======================

With tree-based layout implemented, let's use it in the browser
itself, in its `layout` method. First, we need to create the layout
tree and lay it out:

``` {.python}
class Browser:
    def layout(self, tree):
        document = DocumentLayout(tree)
        document.layout()
```

Once the page is laid out, it must be drawn, which means first
collecting a display list of things to draw and then calling
`self.render()` to actually draw those things. To collect the display
list itself, we'll need to recurse down the tree; I think it's most
convenient to do that by adding a `draw` function to each layout
object which does the recursion.

A neat trick when accumulating a list in a recursive function is to
pass the list itself in as an argument, and have the method just
append to that list instead of returning anything. For
`DocumentLayout`, which only has one child, it looks like this:

``` {.python}
class DocumentLayout:

>     def draw(self, to):

Reviewer: ```suggestion
    def draw(self, display_list):
```

>         self.children[0].draw(to)

Reviewer: ```suggestion
        self.children[0].draw(display_list)
```

```

> For `BlockLayout`, which has multiple, `draw` is called on each child::

Reviewer: ```suggestion
For `BlockLayout`, which has multiple children, `draw` is called on each child::
```

``` {.python}
class BlockLayout:

>     def draw(self, to):

Reviewer: s/to/display_list/ etc.

        # ...
```

Make sure this code come *before* recursively calling `draw` on child
layout objects: the background has to be drawn *before* and therefore
*below* the text inside the source block.

The `render` method now needs to run graphics commands on our actual
canvas. Let's do this with a `draw` method for each command. On
`DrawText` it calls to `create_text`:

``` {.python}
class DrawText:
    def draw(self, scroll, canvas):
        canvas.create_text(
            self.x1, self.y1 - scroll,
            text=self.text,
            font=self.font,
            anchor='nw',
        )
```

Note that the `draw` command accepts the scroll amount as a parameter;
in effect, we're asking each graphics command to perform coordinate
conversion.

`DrawRect` works the same way, except it calls `create_rectangle`. By
default `create_rectangle` draws a one-pixel black border, so make
sure to pass `width = 0`:

``` {.python}

class DrawRect:
    def draw(self, scroll, canvas):
        canvas.create_rectangle(
            self.x1, self.y1 - scroll,
            self.x2, self.y2 - scroll,
            width=0,
            fill=self.color,
        )
```

> We do still want to clip out graphics commands that only occur

Reviewer: ```suggestion
We do still want to omit graphics commands that only occur
```

offscreen. `DrawRect` already contains a `y2` field we can use, so

> let's add the same `DrawText`:

Reviewer: ```suggestion
let's add the same to `DrawText`:
```

``` {.python}
def __init__(self, x1, y1, text, font):
    # ...

>     self.y2 = y1 + font.metrics("linespace")

Reviewer: Rename y2 to a more descriptive name, etc.

```

::: {.quirks}

> On some systems, the `measure` and `metrics` commands are awfully

Reviewer: This is definitely true. Might be interesting to note that real browsers go to a lot of lengths to add extra caches and avoid doing work when not necessary because of this problem.

slow. Adding another call will make things a lot slower.

Luckily, this `metrics` call duplicates a call in `flush`. If you're
careful you can pass the results of that call to `DrawText` as an
argument.
:::

With the drawing logic now inside the drawing commands themselves,
the browser's `render` method just determines which commands
to call `draw` for:

``` {.python}
def render(self):
    self.canvas.delete("all")
    for cmd in self.display_list:
        if cmd.y1 > self.scroll + HEIGHT: continue
        if cmd.y2 < self.scroll: continue
        cmd.draw(self.scroll, self.canvas)
```

> Not only have access to the height of any element, but also of the

Reviewer: ```suggestion
We not only have access to the height of any element, but also of the
```

whole page. We can use that to stop the user from scrolling past the
bottom of the page. In `layout`, store the height in a `max_y`
variable:

``` {.python}
def layout(self, tree):
    # ...

>     self.max_y = document.h

Reviewer: Isn't it height minus viewport height?

    self.scroll = max(0, self.scroll))
    self.render()
```

Make sure those `max` and `min` calls happen in the right order!

Summary
=======

This chapter was a dramatic rewrite of your browser's layout engine.
That means:

- Layout is now tree-based and produces a *layout tree*
- Each node in the tree has one of two different *layout modes*
- Each layout object has a size and position computed
- The display list now contains generic commands
- Backgrounds for source code blocks are now drawn.

Tree-based layout makes it possible to dramatically expand our
browser's styling capabilities. We'll work on that in the [next
chapter](styles.md).

Exercises
=========

*Links Bar*: At the top and bottom of each chapter of this book is a
gray bar naming the chapter and offering back and forward links. It is
enclosed in a `<nav class="links">` tag. Have your browser give this
links bar the same light gray background that any other browser does.

*Hidden Head*: There's a good chance your browser is still showing
scripts, styles, and page titles at the top of every page you visit.
Make it so that the `<head>` element and its contents are never
displayed.

> *Bullets*: Add bullets to list items, which in HTML are `<li>` tags.

Reviewer: This one is probably ok, but display:list-item is yet another beast in reality...
